File: app/Models/Trade.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

#[Fillable([
    'account_id', 'symbol', 'direction', 'lot',
    'entry_price', 'sl_price', 'tp_price', 'exit_price',
    'pnl', 'opened_at', 'closed_at',
    'setup', 'group_id', 'pre_group', 'notes', 'source', 'ai_raw',
])]
class Trade extends Model
{
    use Concerns\HasHashid;

    protected function casts(): array
    {
        return [
            'opened_at' => 'datetime',
            'closed_at' => 'datetime',
            'lot' => 'decimal:2',
            'entry_price' => 'decimal:5',
            'sl_price' => 'decimal:5',
            'tp_price' => 'decimal:5',
            'exit_price' => 'decimal:5',
            'pnl' => 'decimal:2',
            'rr_planned' => 'decimal:2',
            'rr_realized' => 'decimal:2',
            'pre_group' => 'array',
            'ai_raw' => 'array',
        ];
    }

    /**
     * `status`, `rr_planned` dan `rr_realized` selalu diturunkan dari harga &
     * pnl — dipasang di event `saving` supaya semua jalur simpan (form manual,
     * import AI, seeder, tinker) menghasilkan nilai yang sama.
     */
    protected static function booted(): void
    {
        static::saving(function (Trade $trade) {
            $trade->rr_planned = $trade->computeRrPlanned();
            $trade->rr_realized = $trade->computeRrRealized();
            $trade->status = $trade->computeStatus();
        });
    }

    public function account(): BelongsTo
    {
        return $this->belongsTo(Account::class);
    }

    /**
     * Posisi stop loss terhadap entry. Stop yang sudah digeser ke entry atau
     * melewatinya bukan kesalahan input — itu manajemen risiko biasa.
     */
    public const STOP_RISK = 'risk';           // masih di sisi rugi

    public const STOP_BREAKEVEN = 'breakeven'; // persis di harga entry

    public const STOP_LOCKED = 'locked';       // sudah lewat entry, profit terkunci

    public function stopState(): ?string
    {
        if ($this->entry_price === null || $this->sl_price === null) {
            return null;
        }

        $entry = (float) $this->entry_price;
        $sl = (float) $this->sl_price;

        if ($sl === $entry) {
            return self::STOP_BREAKEVEN;
        }

        // Risiko buy ada di bawah entry, risiko sell di atasnya.
        return ($this->sign() === 1 ? $sl < $entry : $sl > $entry)
            ? self::STOP_RISK
            : self::STOP_LOCKED;
    }

    /** +1 untuk buy, -1 untuk sell. */
    public function sign(): int
    {
        return $this->direction === 'sell' ? -1 : 1;
    }

    /**
     * Jarak entry ke stop loss — penyebut semua perhitungan R.
     *
     * Hanya stop yang masih di sisi rugi yang punya arti sebagai risiko. Begitu
     * stop digeser ke break-even atau ke sisi profit, risiko awalnya tidak lagi
     * tercatat di mana pun, jadi R memang tidak bisa dihitung — null, bukan angka
     * karangan dari jarak yang sekarang justru mengunci profit.
     */
    private function risk(): ?float
    {
        return $this->stopState() === self::STOP_RISK
            ? abs((float) $this->entry_price - (float) $this->sl_price)
            : null;
    }

    public function computeRrPlanned(): ?float
    {
        $risk = $this->risk();

        if ($risk === null || $this->tp_price === null) {
            return null;
        }

        return round(abs((float) $this->tp_price - (float) $this->entry_price) / $risk, 2);
    }

    public function computeRrRealized(): ?float
    {
        $risk = $this->risk();

        if ($risk === null || $this->exit_price === null) {
            return null;
        }

        $moved = ((float) $this->exit_price - (float) $this->entry_price) * $this->sign();

        return round($moved / $risk, 2);
    }

    public function computeStatus(): string
    {
        return match (true) {
            (float) $this->pnl > 0 => 'win',
            (float) $this->pnl < 0 => 'loss',
            default => 'be',
        };
    }
}

File: mobile/lib/models/trade.dart
import '../core/format.dart';
import '../core/json.dart';

/// Posisi stop loss terhadap entry — lihat `Trade::stopState()` di server.
enum StopState {
  /// Masih di sisi rugi: R bisa dihitung.
  risk,

  /// Persis di harga entry: risiko nol.
  breakeven,

  /// Sudah melewati entry: profit terkunci (SL+).
  locked;

  static StopState? parse(Object? value) => switch (value) {
    'risk' => risk,
    'breakeven' => breakeven,
    'locked' => locked,
    _ => null,
  };
}

class Trade {
  const Trade({
    required this.id,
    required this.symbol,
    required this.direction,
    required this.status,
    required this.pnl,
    required this.openedAt,
    this.setup,
    this.groupId,
    this.notes,
    this.source = 'manual',
    this.lot,
    this.entryPrice,
    this.slPrice,
    this.tpPrice,
    this.exitPrice,
    this.stopState,
    this.rrPlanned,
    this.rrRealized,
    this.closedAt,
    this.aiRaw,
  });

  /// Dashboard, kalender, dan daftar trade mengirim bentuk yang sedikit
  /// berbeda; kolom yang tidak dikirim dibiarkan null.
  factory Trade.fromJson(Json json) => Trade(
    id: '${json['id']}',
    symbol: '${json['symbol'] ?? ''}',
    direction: '${json['direction'] ?? 'buy'}',
    status: '${json['status'] ?? 'be'}',
    setup: toStringOrNull(json['setup']),
    groupId: toStringOrNull(json['group_id']),
    notes: toStringOrNull(json['notes']),
    source: '${json['source'] ?? 'manual'}',
    lot: toDoubleOrNull(json['lot']),
    entryPrice: toDoubleOrNull(json['entry_price']),
    slPrice: toDoubleOrNull(json['sl_price']),
    tpPrice: toDoubleOrNull(json['tp_price']),
    exitPrice: toDoubleOrNull(json['exit_price']),
    pnl: toDouble(json['pnl']),
    stopState: StopState.parse(json['stop_state']),
    rrPlanned: toDoubleOrNull(json['rr_planned']),
    rrRealized: toDoubleOrNull(json['rr_realized']),
    openedAt: wallTime('${json['opened_at']}'),
    closedAt: wallTimeOrNull(json['closed_at']),
    aiRaw: json['ai_raw'] is Map ? map(json['ai_raw']) : null,
  );

  /// Hash dari server (lihat `App\Support\Hashid`) — bukan angka urut.
  final String id;
  final String symbol;

  /// `buy` atau `sell`.
  final String direction;

  /// `win`, `loss`, atau `be`.
  final String status;
  final String? setup;

  /// Kunci grup: beberapa trade berurutan yang lahir dari satu ide.
  final String? groupId;
  final String? notes;

  /// `manual` atau `ai` (diisi dari screenshot).
  final String source;
  final double? lot;
  final double? entryPrice;
  final double? slPrice;
  final double? tpPrice;
  final double? exitPrice;
  final double pnl;
  final StopState? stopState;
  final double? rrPlanned;
  final double? rrRealized;
  final DateTime openedAt;
  final DateTime? closedAt;
  final Json? aiRaw;

  bool get isBuy => direction == 'buy';

  /// Hari efektif trade: hari ia ditutup, sama seperti `COALESCE(closed_at, opened_at)`.
  DateTime get day => closedAt ?? openedAt;

  String get dayKey => isoDate(day);

  List<String> get setups => (setup ?? '')
      .split(',')
      .map((item) => item.trim())
      .where((item) => item.isNotEmpty)
      .toList();
}

/// Filter `/trades`. Record, supaya dua filter yang isinya sama dianggap sama
/// dan provider-nya tidak memuat ulang.
typedef TradeFilters = ({
  String symbol,
  String setup,
  String q,
  String status,
  String stop,
  String direction,
  String from,
  String to,
});

const TradeFilters noTradeFilters = (
  symbol: '',
  setup: '',
  q: '',
  status: '',
  stop: '',
  direction: '',
  from: '',
  to: '',
);

extension TradeFiltersX on TradeFilters {
  Map<String, dynamic> toQuery() => {
    'symbol': symbol,
    'setup': setup,
    'q': q,
    'status': status,
    'stop': stop,
    'direction': direction,
    'from': from,
    'to': to,
  };

  int get active =>
      toQuery().values.where((value) => '$value'.isNotEmpty).length;
}

/// Satu halaman `/trades`: 25 trade + P/L per hari yang dihitung server,
/// supaya total harian tetap utuh walau harinya terpotong batas halaman.
class TradePage {
  const TradePage({
    required this.items,
    required this.daily,
    required this.symbols,
    required this.setups,
    required this.page,
    required this.lastPage,
    required this.total,
  });

  factory TradePage.fromJson(Json json) {
    final trades = map(json['trades']);

    return TradePage(
      items: list(trades['data'])
          .map((item) => Trade.fromJson(map(item)))
          .toList(),
      daily: map(json['daily'])
          .map((key, value) => MapEntry(key, toDouble(value))),
      symbols: strings(json['symbols']),
      setups: strings(json['setups']),
      page: toInt(trades['current_page']),
      lastPage: toInt(trades['last_page']),
      total: toInt(trades['total']),
    );
  }

  final List<Trade> items;
  final Map<String, double> daily;
  final List<String> symbols;
  final List<String> setups;
  final int page;
  final int lastPage;
  final int total;
}

/// Hasil baca screenshot oleh Gemini, siap mengisi form.
class ExtractResult {
  const ExtractResult({
    required this.data,
    required this.lowConfidence,
    required this.raw,
  });

  factory ExtractResult.fromJson(Json json) => ExtractResult(
    data: map(json['data']),
    lowConfidence: strings(json['low_confidence_fields']),
    raw: json['raw'] is Map ? map(json['raw']) : null,
  );

  final Json data;
  final List<String> lowConfidence;
  final Json? raw;
}
